From 0b5ab7665417fc71a6051cc33a10fc3609401748 Mon Sep 17 00:00:00 2001
From: "ivan.liu" <xiaowen.liu@nxp.com>
Date: Wed, 21 Mar 2018 01:15:02 +0800
Subject: [PATCH 4128/5242] MA-11589 drm: Implement HDR source metadata atomic
 set property.

commit  7478ba3911b8a84306bdbedf3b19629d1c0d184e from
https://source.codeaurora.org/external/imx/linux-imx.git

set_property calls drm_atomic_helper_connector_set_property.
And it will call atomic_set_property which sets metadata.

Change-Id: Ia264cfb34f4744fe3622d2568cdc5f5cee090554
Signed-off-by: ivan.liu <xiaowen.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/drm_atomic.c        |    9 ++++++++-
 drivers/gpu/drm/drm_atomic_helper.c |    4 ++--
 drivers/gpu/drm/imx/hdp/imx-hdp.c   |   12 +++++-------
 3 files changed, 15 insertions(+), 10 deletions(-)

diff --git a/drivers/gpu/drm/drm_atomic.c b/drivers/gpu/drm/drm_atomic.c
index 1a0c950..88ad813 100644
--- a/drivers/gpu/drm/drm_atomic.c
+++ b/drivers/gpu/drm/drm_atomic.c
@@ -1288,9 +1288,16 @@ static int drm_atomic_connector_set_property(struct drm_connector *connector,
 				&state->hdr_source_metadata_blob_ptr,
 				val,
 				-1,
+				-1,
 				&replaced);
 		state->hdr_metadata_changed |= replaced;
-		return ret;
+		if (ret < 0)
+			return ret;
+
+		if (connector->funcs->atomic_set_property) {
+			return connector->funcs->atomic_set_property(connector,
+					state, property, val);
+		}
 	} else if (property == config->aspect_ratio_property) {
 		state->picture_aspect_ratio = val;
 	} else if (property == connector->scaling_mode_property) {
diff --git a/drivers/gpu/drm/drm_atomic_helper.c b/drivers/gpu/drm/drm_atomic_helper.c
index 95e4394..c14a320 100644
--- a/drivers/gpu/drm/drm_atomic_helper.c
+++ b/drivers/gpu/drm/drm_atomic_helper.c
@@ -3683,7 +3683,7 @@ void drm_atomic_helper_connector_reset(struct drm_connector *connector)
 		drm_connector_get(connector);
 	state->commit = NULL;
 	if (state->hdr_source_metadata_blob_ptr)
-		drm_property_reference_blob(state->hdr_source_metadata_blob_ptr);
+		drm_property_blob_get(state->hdr_source_metadata_blob_ptr);
 
 	state->hdr_metadata_changed = false;
 }
@@ -3816,7 +3816,7 @@ struct drm_atomic_state *
 	if (state->commit)
 		drm_crtc_commit_put(state->commit);
 	if (state->hdr_source_metadata_blob_ptr)
-		drm_property_unreference_blob(state->hdr_source_metadata_blob_ptr);
+		drm_property_blob_put(state->hdr_source_metadata_blob_ptr);
 }
 EXPORT_SYMBOL(__drm_atomic_helper_connector_destroy_state);
 
diff --git a/drivers/gpu/drm/imx/hdp/imx-hdp.c b/drivers/gpu/drm/imx/hdp/imx-hdp.c
index 5f9797b..9e7a921 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdp.c
+++ b/drivers/gpu/drm/imx/hdp/imx-hdp.c
@@ -677,22 +677,20 @@ static void imx_hdp_connector_force(struct drm_connector *connector)
 }
 
 static int imx_hdp_set_property(struct drm_connector *connector,
+				struct drm_connector_state *state,
 				struct drm_property *property, uint64_t val)
 {
 	struct imx_hdp *hdp = container_of(connector, struct imx_hdp,
 					   connector);
 	int ret;
-	struct drm_connector_state *conn_state;
 	union hdmi_infoframe frame;
 	struct hdr_static_metadata *hdr_metadata;
 
-	conn_state = connector->state;
-
-	if (conn_state->hdr_source_metadata_blob_ptr &&
-	    conn_state->hdr_source_metadata_blob_ptr->length &&
+	if (state->hdr_source_metadata_blob_ptr &&
+	    state->hdr_source_metadata_blob_ptr->length &&
 	    hdp->ops->write_hdr_metadata) {
 		hdr_metadata = (struct hdr_static_metadata *)
-				conn_state->hdr_source_metadata_blob_ptr->data;
+				state->hdr_source_metadata_blob_ptr->data;
 
 		ret = drm_hdmi_infoframe_set_hdr_metadata(&frame.drm,
 							  hdr_metadata);
@@ -716,7 +714,7 @@ static int imx_hdp_set_property(struct drm_connector *connector,
 	.reset = drm_atomic_helper_connector_reset,
 	.atomic_duplicate_state = drm_atomic_helper_connector_duplicate_state,
 	.atomic_destroy_state = drm_atomic_helper_connector_destroy_state,
-	.set_property = imx_hdp_set_property,
+	.atomic_set_property = imx_hdp_set_property,
 };
 
 static const struct drm_connector_helper_funcs imx_hdp_connector_helper_funcs = {
-- 
1.7.9.5

